app/controllers/spree/affirm_controller.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  class AffirmController < Spree::StoreController
+    protect_from_forgery except: [:confirm]
+
+    def confirm
+      order = current_order || raise(ActiveRecord::RecordNotFound)
+      checkout_token = affirm_params[:checkout_token]
+
+      if !checkout_token
+        return redirect_to checkout_state_path(order.state), notice: "Invalid order confirmation data passed in"
+      end
+
+      if order.complete?
+        return redirect_to spree.order_path(order), notice: "Order is already in complete state"
+      end
+
+      affirm_checkout = Solidus::Affirm::Checkout.new(token: checkout_token)
+
+      if affirm_checkout.valid?
+        affirm_checkout.save
+        order.payments.create!({
+          payment_method_id: affirm_params[:payment_method_id],
+          amount: order.total,
+          source: affirm_checkout
+        })
+        order.next
```

Comment:
Some sites may not want to "order.next", what happens if sites use a custom state machine, or don't use the state machine at all? How would they customize this controller in that scenario without implementing their own.